ultrasonic/src/main/kotlin/org/moire/ultrasonic/service/MediaPlayerService.kt

```diff
@@ -816,6 +921,32 @@ class MediaPlayerService : Service() {
                 Timber.v("Media Session Callback: onPlay")
             }
 
+            override fun onPlayFromMediaId(mediaId: String?, extras: Bundle?) {
+                super.onPlayFromMediaId(mediaId, extras)
+
```

Comment:
This here is the last piece of magic. Once a user has selected a song from the list, this function will be called. The `extras` object contains all the data I need to set up the download and play.

Took me forever to find this bit, turns out it's documented right [here](https://developer.android.com/training/cars/media#playback-commands). A lesson to read all the way to the bottom of even long documents :)